### PR TITLE
Add UI Window based on X11

### PR DESCRIPTION
**Pull request by ihhub**
Add a UI class based on X11 technology for Linux systems. Few notes to consider:
1. No line/circle/rectangle drawings in the code
2. Image has fixed size and it doesn't depend on window size
3. No example

### CODE REVIEW

File: src/ui/x11/x11_ui.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+#include <vector>
+#include <X11/Xlib.h>
+#include "../ui.h"
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/ui/x11/x11_ui.h]: `ui.h` file locates in parent directory.

File: src/ui/x11/x11_ui.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
+#include "x11_ui.h"
+#include "../../image_exception.h"
+
+UiWindowX11::UiWindowX11( const PenguinV_Image::Image & image, const std::string & title )
+    : UiWindow( image, title )
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/ui/x11/x11_ui.cpp]: `_data` is a vector. It has it's own default constructor.

File: src/ui/x11/x11_ui.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+UiWindowX11::~UiWindowX11()
+{
+    XDestroyWindow( _uiDisplay, _window );
+    XCloseDisplay( _uiDisplay );
+}
+
+void UiWindowX11::_display()
+{
+    if ( !_shown ) // we don't need to display anything
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/ui/x11/x11_ui.cpp]: It is a member of base class.

File: src/ui/x11/x11_ui.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        return;
+
+    XEvent e;
+    while ( 1 ) {
+        XNextEvent( _uiDisplay, &e );
+        if ( e.type == Expose )
+            XPutImage( _uiDisplay, _window, DefaultGC( _uiDisplay, _screen ), _image, 0, 0, 0, 0, _width, _height );
+        else if ( (e.type == ClientMessage) && (static_cast<unsigned int>(e.xclient.data.l[0]) == _deleteWindowEvent) )
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/ui/x11/x11_ui.cpp]: According to `X.h` file **Atom** is unsigned int:
```cpp
...
typedef unsigned long Atom;		/* Also in Xdefs.h */
...
typedef CARD32 Atom;
..
```
where **CARD32** is unsigned int as well.

File: src/ui/x11/x11_ui.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
+#include "x11_ui.h"
+#include "../../image_exception.h"
+
+UiWindowX11::UiWindowX11( const PenguinV_Image::Image & image, const std::string & title )
+    : UiWindow( image, title )
+    , _uiDisplay( nullptr )
+    , _screen( 0 )
+    , _window( 0 )
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/ui/x11/x11_ui.cpp]: According `X.h` file:
```cpp
...
typedef unsigned long XID;
...
typedef CARD32 XID;
...
typedef XID Window;
```
where **CARD32** is unsigned int.

**Comment by ihhub** [comment]
Hi @yannlabou , I would need your help to create CMake file as it requires to verify an existence of X11 package of machine before compilation of the example. I think we might make such change as a separate issue :)

**Comment by yannlabou** [comment]
> Hi @yannlabou , I would need your help to create CMake file as it requires to verify an existence of X11 package of machine before compilation of the example. I think we might make such change as a separate issue :)

Thanks CMake there is a [FindX11 module](https://cmake.org/cmake/help/latest/module/FindX11.html). Please file a new issue and assign it to me. I might have time next weekend :).

**Comment by ihhub** [comment]
I'm thinking to merge edge detection examples for Windows and Linux since the only difference is WindowUI type. We need this to avoid code duplication.

**Pull request merged**